### ISSUE TITLE
Deriving JsonSchema seems to break serde's "with" field annotation

### ISSUE THREAD

**Issue by joel-wright** [GREsau/schemars#89]
I have been using `#[derive(JsonSchema)]`, but recently ran into issues attempting to use serde's field attributes and the `serde_with` crate in order to report an error on duplicate keys in a mapping.

When I add the field attribute to the struct:

```
#[derive(JsonSchema, Serialize, Deserialize, Debug, PartialEq)]
#[serde(deny_unknown_fields)]
pub(crate) struct MyStruct {
    ...
    #[serde(with = "::serde_with::rust::maps_duplicate_key_is_error")]
    pub mapping: BTreeMap<String, String>,
    ...
}
```

I get the following compilation error:

```
error[E0573]: expected type, found module `serde_with::rust::maps_duplicate_key_is_error`
  --> data/structs.rs:24:20
   |
24 |     #[serde(with = "::serde_with::rust::maps_duplicate_key_is_error")]
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a type

error: aborting due to previous error
```

If I remove `JsonSchema` from the list of derives it seems to work fine.

**Comment by Igosuki** [comment]
I also have trouble with remotes for some specific data structures, seems to have to do with some resolution order. 
And I also have a recursive enum that breaks schema generation with a stackoverflow :)

**Comment by Igosuki** [comment]
@uttarayan21 sounds like something should get patched in the code though

**Comment by uttarayan21** [comment]
Definitely. I just added it here so that people can compile in the meantime.

**Comment by abizjak** [comment]
I came across this today as well and while I think the error when this happens should (and easily can) be better and point to the way things should be done, but there is a straightforward way to use `with` and this might even be the intended way.

For the original example the following would be the solution
```rust
#[derive(JsonSchema, Serialize, Deserialize, Debug, PartialEq)]
#[serde(deny_unknown_fields)]
pub(crate) struct MyStruct {
    ...
    #[serde(with = "::serde_with::rust::maps_duplicate_key_is_error")]
    #[schemars(with = BTreeMap<String, String>)]
    pub mapping: BTreeMap<String, String>,
    ...
}
```
or ideally a type that more specifically does the "no duplicate keys validation" and records that in the schema.

Generally the allowed attributes and their semantics need to be better documented, like they are for serde since I also found it hard to figure out what is and what is not possible.

In my code I've settled on the pattern 

```rust
#[derive(JsonSchema, Serialize, Deserialize, Debug, PartialEq)]
pub(crate) struct MyStruct {
    #[serde(with = "foo")]
    #[schemars(with = MyStructSchema)]
    pub mapping: SomeType,
}

impl JsonSchema for SomeType {
  ...
}
```
in the cases where I needed custom and precise schemas.

**Comment by wackazong** [comment]
I can confirm the bug and the workaround by @uttarayan21 🙏

**Comment by JakkuSakura** [comment]
This seems not hard to fix at first glance, but with #[serde(with)] one can do crazy things.
I guess it's the reason this hasn't been picked up for years

**Comment by GREsau** [comment]
This occurs because the `#[serde(with = "...")]` attribute is respected by schemars in order to enable [remote deriving](https://graham.cool/schemars/examples/5-remote_derive/). But as you've discovered, this breaks when the `with` value is not a type that implements `JsonSchema`.

The easiest way to resolve this would be using @abizjak's workaround, although you must include quotes around the `with` value like so:
```rust
#[derive(JsonSchema, Serialize, Deserialize, Debug, PartialEq)]
#[serde(deny_unknown_fields)]
pub(crate) struct MyStruct {
    #[serde(with = "::serde_with::rust::maps_duplicate_key_is_error")]
    #[schemars(with = "BTreeMap<String, String>")]
    pub mapping: BTreeMap<String, String>,
}
```

I'm afraid I don't know how this could be reliably fixed within schemars, because (as far as I know) the proc macro has no way of knowing whether the `with` value (in this case `::serde_with::rust::maps_duplicate_key_is_error`) refers to a module or a type.

**Comment by oxalica** [comment]
> I'm afraid I don't know how this could be reliably fixed within schemars, because (as far as I know) the proc macro has no way of knowing whether the `with` value (in this case `::serde_with::rust::maps_duplicate_key_is_error`) refers to a module or a type.

Why is it necessary to accept both types and module paths? According to docs of serde, they always accept a "path", not a type, and that's the reason why they require `#[serde(with = "Vec::<DurationDef>")]` in [remote deriving](https://graham.cool/schemars/examples/5-remote_derive/).

Is there any reason prevents us from doing the same to always parse `serde(with = ...)` to a path? We could require the same for `schemars(with = ...)` for consistency, but can also make them accept different syntax for compatibility.

**Comment by GREsau** [comment]
@oxalica So you're suggesting that, rather than requiring the `with` param to be a type that implements `JsonSchema`, schemars also allows it to be a module? In order for that to work, the module would have to expose a compatible `json_schema` function at a minimum, and probably also `always_inline_schema`, `schema_name` and `schema_id` functions. That's possible, but I don't think it really solves the original problem in this issue

**Comment by oxalica** [comment]
> @oxalica So you're suggesting that, rather than requiring the `with` param to be a type that implements `JsonSchema`, schemars also allows it to be a module? In order for that to work, the module would have to expose a compatible `json_schema` function at a minimum, and probably also `always_inline_schema`, `schema_name` and `schema_id` functions. That's possible, but I don't think it really solves the original problem in this issue

Sorry, I think I misunderstood the issue. I was suggesting to use https://github.com/GREsau/schemars/issues/89#issuecomment-2284582188 as the default behavior, that is, parse `serde(with = ...)` as serde does and ignore it, and use the field definition type for schema generation, unless `schemars(with = ...)` is also given.

But yes, it's a trade off to make validation or preprocessing `serde(with = ...)` more convenient, with the cost of more code for remote derive. We can also use a compatible parser for `serde` attributes in this way, so adding optional `schemars` support for a library would not force a change to unrelated `serde(..)` attributes.
